subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/VersionConflictResolutionIntegrationTest.groovy

```diff
@@ -1781,4 +1781,66 @@ dependencies {
         false           | false
     }
 
+    @Issue("gradle/gradle#8944")
+    def 'verify that cleaning up constraints no longer causes a ConcurrentModificationException'() {
```

Comment:
Nice test :+1:

Reply:
Yup, learned that in some collections, if you purge the collection while iterating it, you will not hit a `ConcurrentModificationException` because `hasNext()` does not check for modifications! If it were, we would have seen this problem a long time ago.